File: Python/ByTier/C/1767_C_Count_Binary_Strings.py
# Problem: CF 1767 C - Count Binary Strings
# https://codeforces.com/contest/1767/problem/C

"""
Count Binary Strings

Algorithm:
This problem is solved using dynamic programming. The idea is to build the binary string from left to right, keeping track of the number of valid strings that end at each position with a specific character. 

We define dp[i][j] as the number of valid ways to fill the first i positions of the string such that the i-th character is j (where j=0 or j=1).

The constraints are defined for all substrings [i,j] with a[i,j] which can be:
- 1: all characters in substring must be the same
- 2: at least two different characters in substring
- 0: no constraint

Time Complexity: O(n^3)
Space Complexity: O(n^2)

Approach:
1. Parse input matrix a[i][j] which represents constraints for substring [i,j]
2. Use dynamic programming where dp[i][j] stores number of valid strings of length i ending with character j
3. For each position, update dp values based on constraint checks
4. Sum all dp[n][j] for final answer

"""

n = int(input())
N = 105
a = [0] * N
for i in range(1, n + 1):
    a[i] = [0] * i + list(map(int, input().split()))
f = [[0] * N for i in range(N)]
mod = 998244353
f[1][1] = 2  # Initially we have 2 ways to fill first position (0 or 1)

# Iterate through all positions of the string
for i in range(1, n + 1):
    for j in range(1, i + 1):
        # Check if current configuration is valid given constraints
        flag = True
        for k in range(1, i + 1):
            if a[k][i] == 1 and k < j:
                flag = False  # If any 1 constraint (all same) is violated
            if a[k][i] == 2 and k >= j:
                flag = False  # If any 2 constraint (at least two different) is violated
        # If constraints are satisfied, accumulate the number of valid sequences
        if flag == False:
            f[i][j] = 0
        f[i + 1][j] = (f[i + 1][j] + f[i][j]) % mod  # Propagate valid sequences to next position with same ending character
        f[i + 1][i + 1] = (f[i + 1][i + 1] + f[i][j]) % mod  # Propagate to new character (i+1) at position i+1

res = 0
# Sum up all valid ways to finish at position n
for i in range(1, n + 1):
    res = (res + f[n][i]) % mod
print(res)


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1767/1767_C_Count_Binary_Strings.cpp
// Problem: CF 1767 C - Count Binary Strings
// https://codeforces.com/contest/1767/problem/C

/*
 * Problem: Count Binary Strings
 * Algorithm: Dynamic Programming with Segment Checking
 * 
 * Time Complexity: O(n^4)
 * Space Complexity: O(n^2)
 * 
 * Approach:
 * - Use dynamic programming where dp[i] represents the number of valid
 *   ways to fill the first i positions of the string.
 * - For each position i, check all previous positions j (0 to i-1) to see
 *   if we can transition from dp[j] to dp[i] by appending a new segment
 *   from position j+1 to i.
 * - The check function validates whether the segment [j+1, i] adheres to
 *   the constraints defined in the input matrix s.
 * - Since we can start with either '0' or '1', the final answer is multiplied by 2.
 */

#include <stdio.h>
#include <iosfwd>

using namespace std;
const int p = 998244353;
int n, s[105][105];
int dp[105];

// Check if segment [l, r] satisfies constraints
bool check(int l, int r) {
  for (int i = 1; i <= r; i++)
    for (int j = i; j <= n; j++) {
      // If constraint is 1 (all same), but segment overlaps with [l, r] and it's not fully contained
      if (i < l && l <= j && s[i][j] == 1)
        return false;
      // If constraint is 2 (at least two different chars), but segment overlaps with [l, r] and it's not fully contained
      if (l <= i && j <= r && s[i][j] == 2)
        return false;
    }
  return true;
}

int main() {
  scanf("%d", &n);
  for (int i = 1; i <= n; i++)
    for (int j = i; j <= n; j++)
      scanf("%d", &s[i][j]);
  
  // Base case: empty prefix has one way
  dp[0] = 1;
  
  // Fill dp array
  for (int i = 1; i <= n; i++)
    for (int j = 0; j < i; j++)
      // If it's valid to place segment [j+1, i], add ways
      if (check(j + 1, i))
        dp[i] = (dp[i] + dp[j]) % p;
  
  // Multiply by 2 because we can start with either 0 or 1
  printf("%d", dp[n] * 2 % p);
}


// https://github.com/VaHiX/CodeForces/